File: scripts/quickstart-code-samples/recommendTools.ts
/**
 * AgentRank Quickstart — TypeScript
 * Recommend MCP tools for a given use case using the AgentRank v1 API.
 *
 * No API key required for basic search.
 * Get a free v2 API key at https://agentrank-ai.com/docs
 *
 * Usage:
 *   npx tsx recommendTools.ts
 *   npx tsx recommendTools.ts "query postgres database"
 *
 * Requires: Node.js 18+ (built-in fetch) or npx tsx for TypeScript execution
 */

const BASE_URL = "https://agentrank-ai.com/api/v1/search";

interface AgentRankResult {
  type: "tool" | "skill" | "agent";
  id: string;
  name: string;
  description: string | null;
  score: number;
  rank: number;
  url: string;
  raw: Record<string, unknown>;
}

interface SearchResponse {
  query: string | null;
  category: string;
  sort: string;
  results: AgentRankResult[];
  meta: { total: number; limit: number; offset: number };
}

/**
 * Search the AgentRank index.
 *
 * @param query - Keyword search, e.g. "database" or "browser automation"
 * @param category - "tool" (GitHub repos), "skill" (registry entries), or "" (all)
 * @param sort - "score" (default), "rank", "stars", or "freshness"
 * @param limit - Number of results (max 100)
 */
export async function searchTools(
  query: string,
  category: "tool" | "skill" | "agent" | "" = "tool",
  sort: "score" | "rank" | "stars" | "freshness" = "score",
  limit = 10
): Promise<AgentRankResult[]> {
  const url = new URL(BASE_URL);
  url.searchParams.set("q", query);
  if (category) url.searchParams.set("category", category);
  url.searchParams.set("sort", sort);
  url.searchParams.set("limit", String(limit));

  const res = await fetch(url.toString());
  if (!res.ok) throw new Error(`AgentRank API error: ${res.status} ${res.statusText}`);
  const data: SearchResponse = await res.json();
  return data.results;
}

/**
 * Return top-ranked MCP tools for a given use case.
 *
 * @param useCase - Natural-language description, e.g. "scrape web pages"
 * @param topN - Number of recommendations (default 5)
 * @returns Formatted string suitable for including in an LLM prompt
 */
export async function recommendTools(useCase: string, topN = 5): Promise<string> {
  const results = await searchTools(useCase, "tool", "score", topN);

  if (!results.length) {
    return `No ranked tools found for '${useCase}'. Check agentrank-ai.com for the latest index.`;
  }

  const verdict = (score: number) =>
    score >= 80 ? "highly rated" : score >= 60 ? "solid" : "use with caution";

  const lines = [`Top ${results.length} MCP tools for '${useCase}' (ranked by AgentRank score):\n`];
  results.forEach((tool, i) => {
    lines.push(`${i + 1}. ${tool.name} — score ${tool.score.toFixed(1)} (${verdict(tool.score)}, #${tool.rank})`);
    if (tool.description) lines.push(`   ${tool.description}`);
    lines.push(`   ${tool.url}`);
  });
  return lines.join("\n");
}

/**
 * Fetch full score breakdown for a specific tool (requires v2 API key).
 *
 * @param githubFullName - e.g. "modelcontextprotocol/python-sdk"
 * @param apiKey - Your AgentRank API key (get free at agentrank-ai.com/docs)
 */
export async function getToolDetails(githubFullName: string, apiKey: string): Promise<unknown> {
  const toolId = githubFullName.replace("/", "--");
  const res = await fetch(`https://agentrank-ai.com/api/v2/tool/${toolId}`, {
    headers: { Authorization: `Bearer ${apiKey}` },
  });
  if (!res.ok) throw new Error(`AgentRank API error: ${res.status}`);
  return res.json();
}

// Run standalone
const useCase = process.argv.slice(2).join(" ") || "scrape web pages";
console.log(await recommendTools(useCase));
console.log();
console.log(await recommendTools("query postgres database"));

File: scripts/quickstart-code-samples/recommend_tools.py
#!/usr/bin/env python3
"""
AgentRank Quickstart — Python
Recommend MCP tools for a given use case using the AgentRank v1 API.

No API key required for basic search.
Get a free v2 API key at https://agentrank-ai.com/docs

Usage:
    python recommend_tools.py
    python recommend_tools.py "query postgres database"
"""

import sys
import requests

BASE_URL = "https://agentrank-ai.com/api/v1/search"


def search_tools(query: str, category: str = "tool", sort: str = "score", limit: int = 10) -> list[dict]:
    """Search the AgentRank index.

    Args:
        query: Keyword search, e.g. "database" or "browser automation"
        category: "tool" (GitHub repos), "skill" (registry entries), or "" (all)
        sort: "score" (default), "rank", "stars", or "freshness"
        limit: Number of results (max 100)

    Returns:
        List of result dicts with type, name, score, rank, url, raw fields.
    """
    resp = requests.get(
        BASE_URL,
        params={"q": query, "category": category, "sort": sort, "limit": limit},
        timeout=10,
    )
    resp.raise_for_status()
    return resp.json()["results"]


def recommend_tools(use_case: str, top_n: int = 5) -> str:
    """Return top-ranked MCP tools for a given use case.

    Args:
        use_case: Natural-language description, e.g. "scrape web pages"
        top_n: Number of recommendations to return (default 5)

    Returns:
        A formatted string suitable for including in an LLM prompt.
    """
    results = search_tools(use_case, limit=top_n)

    if not results:
        return f"No ranked tools found for '{use_case}'. Check agentrank-ai.com for the latest index."

    def verdict(score: float) -> str:
        if score >= 80:
            return "highly rated"
        if score >= 60:
            return "solid"
        return "use with caution"

    lines = [f"Top {len(results)} MCP tools for '{use_case}' (ranked by AgentRank score):\n"]
    for i, tool in enumerate(results, 1):
        score = tool["score"]
        lines.append(f"{i}. {tool['name']} — score {score:.1f} ({verdict(score)}, #{tool['rank']})")
        if tool.get("description"):
            lines.append(f"   {tool['description']}")
        lines.append(f"   {tool['url']}")
    return "\n".join(lines)


def get_tool_details(github_full_name: str, api_key: str) -> dict:
    """Fetch full score breakdown for a specific tool (requires v2 API key).

    Args:
        github_full_name: e.g. "modelcontextprotocol/python-sdk"
        api_key: Your AgentRank API key (get free at agentrank-ai.com/docs)

    Returns:
        Tool detail dict including signals, rankHistory, and metadata.
    """
    tool_id = github_full_name.replace("/", "--")
    resp = requests.get(
        f"https://agentrank-ai.com/api/v2/tool/{tool_id}",
        headers={"Authorization": f"Bearer {api_key}"},
        timeout=10,
    )
    resp.raise_for_status()
    return resp.json()


if __name__ == "__main__":
    use_case = " ".join(sys.argv[1:]) if len(sys.argv) > 1 else "scrape web pages"
    print(recommend_tools(use_case))
    print()
    print(recommend_tools("query postgres database"))
